libs/synchronization/include/hpx/synchronization/latch.hpp

```diff
@@ -128,28 +142,36 @@ namespace hpx { namespace lcos { namespace local {
         {
             HPX_ASSERT(update >= 0);
 
-            std::unique_lock<mutex_type> l(mtx_.data_);
-
             std::ptrdiff_t new_count = (counter_ -= update);
             HPX_ASSERT(new_count >= 0);
 
             if (new_count == 0)
             {
-                notified_ = true;
-                cond_.data_.notify_all(std::move(l));    // release the threads
+                bool expected = false;
+                if (notified_.compare_exchange_strong(
+                        expected, true, std::memory_order_acq_rel))
+                {
+                    // release the threads
+                    std::unique_lock<mutex_type> l(mtx_.data_);
+                    cond_.data_.notify_all(std::move(l));
+                }
```

Comment:
This looks correct to me. My only question for the above block is: would `notified_` ever be `true` before a thread has decremented the count to 0? The thread that decrements the count to 0 is always the one that notifies as well, right? That is, should we assert that the `compare_exchange_strong` always returns `true`?

Reply:
@msimberg Hmmm. I think you're right. There is no way for the `notified_` flag to be `true` without the `counter_` reaching zero first, which can happen on the same thread of execution only. I changed the implementation accordingly, please have another look.

Reply:
Now the new assertion fires: https://cdash.cscs.ch/testDetails.php?test=26951590&build=124891. I guess that means that we should keep the `compare_exchange`. :/

Reply:
Or that the latch being used somehow inappropriately there? I'd like to at least check what the full stack trace for that failure is before we merge this. I'll try to do this today.

Reply:
The other segmentation faults also don't look very promising :/

Reply:
Thanks! The changes here are not very important, we can drop them on the floor without losing anything.